### PR TITLE
gpt-auto-generator: mount ESP when LoaderDevicePartUUID isn't set

### PR DESCRIPTION
**Pull request by dbnicholson**
If the boot loader doesn't set LoaderDevicePartUUID, then also add the ESP automount if it's on the same device as the root partition. This is a weaker guarantee but likely the same for the vast majority of systems. Allowing the ESP automount in this case helps break a dependency loop. Existing boot loaders can be changed to set LoaderDevicePartUUID, but they can't be delivered to existing systems if the ESP is not mounted.

### CODE REVIEW

**Comment by poettering** [comment]
wouldn't it make more sense to just teach that boot loader you are using to set the efi var? this will improve things for everyone.

File: src/gpt-auto-generator/gpt-auto-generator.c [hunk 1]
```diff
@@ -734,7 +737,8 @@ static int enumerate_partitions(dev_t devnum) {
         }
 
         if (m->partitions[PARTITION_ESP].found) {
-                k = add_partition_esp(m->partitions + PARTITION_ESP, m->partitions[PARTITION_XBOOTLDR].found);
+                k = add_partition_esp(m->partitions + PARTITION_ESP, m->partitions[PARTITION_XBOOTLDR].found,
+                                      m->partitions[PARTITION_ROOT].found);
```
Comment (poettering) [inline src/gpt-auto-generator/gpt-auto-generator.c]: this is kinda pointless. when called after the initrd transition gpt-auto-generator already operates on the device anyway that backs the current root device.

moreover this check you added here actually checks for "hey, does this block device have a properly tagged root partition", which is quite different from "does this block device have the root partition we are currently using on it?", which your commit msg suggests you are looking for (but which is pointless, as mentioned).

**Comment by poettering** [comment]
i am not sure what the goal of this patch is supposed to be? workaround for legacy boot loaders with problematic upstreams?

File: src/gpt-auto-generator/gpt-auto-generator.c (hunk 1 above)
Comment (dbnicholson) [reply src/gpt-auto-generator/gpt-auto-generator.c]: Oh, I misunderstood that API. Fair enough.

**Comment by dbnicholson** [comment]
> i am not sure what the goal of this patch is supposed to be? workaround for legacy boot loaders with problematic upstreams?

Allow us to mount the ESP so that we can get an updated boot loader to existing systems. We use systemd-boot for some systems, but grub for most. Even if grub did start setting the EFI variable, we can't deploy that updated grub to existing systems if they don't have the ESP mounted. We can manage this downstream, though.

**Comment by keszybz** [comment]
So… I can see how this is useful in places where grub2 is deployed. The patches that Oliver posted didn't even get a single reply, so it's likely that this won't be widely available anytime soon.

But there's a real risk from removing this check. We set `LoaderDevicePartUUID` in two places: `sd-boot` and `sd-stub`. In both cases the setting is conditional. But we also unconditionally try to set `LoaderFirmwareInfo` and `LoaderFirmwareType`. To make this safer, I'd propose the following modification: check if `LoaderFirmwareInfo` is set. If it is, treat missing `LoaderDevicePartUUID` as before. But if `LoaderDevicePartUUID` is also not set, assume that the boot loader does not support BLI and do what your current patch does.

@poettering WDYT?

**Comment by dbnicholson** [comment]
> So… I can see how this is useful in places where grub2 is deployed. The patches that Oliver posted didn't even get a single reply, so it's likely that this won't be widely available anytime soon.

The maintainer did actually reply but it came the following month so doesn't show up on that page. It does seem like it's getting somewhere. Here's the latest series and the maintainer's latest reply:

https://lists.gnu.org/archive/html/grub-devel/2023-03/msg00120.html
https://lists.gnu.org/archive/html/grub-devel/2023-04/msg00008.html

**Comment by poettering** [comment]
> https://lists.gnu.org/archive/html/grub-devel/2023-03/msg00120.html
> https://lists.gnu.org/archive/html/grub-devel/2023-04/msg00008.html

any further update to this? that mail doesn't look particularly negative to me

**Comment by poettering** [comment]
So I am concerned about this, because it basically means that if you have *any* disk connected to your system with an ESP we might mount it, and since we also derive the root fs from it even the rootfs. Even if it's a removable usb stick. What pops up first might win. That's entirely arbitrary. The efi var we added for a reason: so that there's not doubt where to look for a root partition and an ESP.

This is not just a security issue but also a robustness issue: any auto-detection of the rootfs and ESP is useless if it comes to completely random results if you happen to plug in a usb disk to your system...

**Comment by poettering** [comment]
Hmm, so I found this:

https://lists.gnu.org/archive/html/grub-devel/2023-05/msg00156.html

does this mean it got merged?

**Comment by poettering** [comment]
Seems it was:

https://git.savannah.gnu.org/cgit/grub.git/tree/grub-core/commands/bli.c

Perfect then. That's a much better outcome. Let's close this PR then, as this doesn't seem necessary anymore (and I have major reservations regarding the logic behind it anway).

**Pull request closed without merging**